### PR TITLE
[Gateway] Add `min_stake` gateway module param

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Adds the `min_stake_gateway` param to the gateway.

## Depends on

- #808 
- #827
- #826

## Issue

- #612

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [x] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -223,6 +223,9 @@ genesis:
             amount: "1000068"
             denom: upokt
     gateway:
+      params:
+        # TODO_MAINNET: Determine realistic amount for minimum gateway stake amount.
+        min_stake: 100
```
Comment (red-0ne) [inline config.yml:228]: This should be a coin.

File: tools/scripts/params/gateway_min_stake.json [hunk 2]
```diff
@@ -2,9 +2,9 @@
   "body": {
     "messages": [
       {
-        "@type": "/poktroll.tokenomics.MsgUpdateParam",
+        "@type": "/poktroll.gateway.MsgUpdateParam",
         "authority": "pokt10d07y265gmmuvt4z0w9aw880jnsr700j8yv32t",
-        "name": "compute_units_to_tokens_multiplier",
+        "name": "min_stake",
         "as_int64": "42"
```
Comment (red-0ne) [inline tools/scripts/params/gateway_min_stake.json:8]: Should be `as_coin`

File: x/gateway/keeper/msg_server_update_param.go [hunk 3]
```diff
@@ -3,15 +3,45 @@ package keeper
...
+		}
+
+		params.MinStake = value.AsCoin
+	default:
+		return nil, types.ErrGatewayParamInvalid.Wrapf("unsupported param %q", msg.Name)
+	}
+
+	// Perform a global validation on all params, which includes the updated param.
+	// This is needed to ensure that the updated param is valid in the context of all other params.
+	if err := params.Validate(); err != nil {
```
Comment (red-0ne) [inline x/gateway/keeper/msg_server_update_param.go:34]:
```suggestion
	if err := params.ValidateBasic(); err != nil {
```

File: x/gateway/keeper/msg_server_update_param_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+	// Set the parameters to their default values
+	k, msgSrv, ctx := setupMsgServer(t)
+	defaultParams := gatewaytypes.DefaultParams()
+	require.NoError(t, k.SetParams(ctx, defaultParams))
+
+	// Ensure the default values are different from the new values we want to set
+	require.NotEqual(t, expectedMinStake, defaultParams.MinStake)
+
+	// Update the min relay difficulty bits
```
Comment (red-0ne) [inline x/gateway/keeper/msg_server_update_param_test.go:27]:
```suggestion
	// Update the min stake
```

File: x/gateway/keeper/params_test.go [hunk 5]
```diff
@@ -16,3 +20,41 @@ func TestGetParams(t *testing.T) {
 	require.NoError(t, k.SetParams(ctx, params))
 	require.EqualValues(t, params, k.GetParams(ctx))
 }
+
+func TestParams_ValidateMinRelayDifficulty(t *testing.T) {
```
Comment (red-0ne) [inline x/gateway/keeper/params_test.go:24]:
```suggestion
func TestParams_ValidateMinStake(t *testing.T) {
```

File: x/gateway/keeper/params_test.go [hunk 6]
```diff
@@ -16,3 +20,41 @@ func TestGetParams(t *testing.T) {
...
+		minStake    any
+		expectedErr error
+	}{
+		{
+			desc:        "invalid type",
+			minStake:    int64(-1),
+			expectedErr: gatewaytypes.ErrGatewayParamInvalid.Wrapf("invalid type for %s: int64; expected *cosmostypes.Coin", gatewaytypes.ParamMinStake),
+		},
+		{
+			desc: "MinStake less than zero",
```
Comment (red-0ne) [inline x/gateway/keeper/params_test.go:36]: Shouldn't we validate that it's greater than zero?

**Review by red-0ne — CHANGES_REQUESTED** [review]
Left a few change requests but looks good overall.

File: x/gateway/types/params.go [hunk 7]
```diff
@@ -1,30 +1,88 @@
...
+		return ErrGatewayParamInvalid.Wrap("missing min_stake")
+	}
+
+	if err := ValidateMinStakeDenom(minStakeCoin); err != nil {
+		return err
+	}
+
+	if err := ValidateMinStakeAboveZero(minStakeCoin); err != nil {
+		return err
+	}
```
Comment (bryanchriswhite) [inline x/gateway/types/params.go:71]:
```suggestion
	if !minStakeCoin.IsValid() {
		return ErrGatewayParamInvalid.Wrapf("invalid min_stake %s; (%v)", minStakeCoin, minStakeCoin.Validate())
	}

	if minStakeCoin.IsZero() || minStakeCoin.IsNegative() {
		return ErrGatewayParamInvalid.Wrapf("invalid min_stake amount: %s <= 0", minStakeCoin)
	}

	if minStakeCoin.Denom != volatile.DenomuPOKT {
		return ErrGatewayParamInvalid.Wrapf("invalid min_stake amount denom %s", minStakeCoin)
	}

```

File: x/gateway/types/params.go [hunk 8]
```diff
@@ -1,30 +1,88 @@
...
+	if minStakeCoin.Denom != volatile.DenomuPOKT {
+		return ErrGatewayParamInvalid.Wrapf("min stake denom must be %s: %s", volatile.DenomuPOKT, minStakeCoin)
+	}
+	return nil
+}
+
+func ValidateMinStakeAboveZero(minStakeCoin *cosmostypes.Coin) error {
+	if minStakeCoin.Amount.Int64() <= 0 {
+		return ErrGatewayParamInvalid.Wrapf("min stake amount must be greater than zero: %s", minStakeCoin)
+	}
```
Comment (bryanchriswhite) [inline x/gateway/types/params.go:86]:
```suggestion
```

File: config.yml (hunk 1 above)
Comment (bryanchriswhite) [reply config.yml:228]: Nice catch! :sunglasses:

File: x/gateway/keeper/msg_server_update_param.go (hunk 3 above)
Comment (bryanchriswhite) [reply x/gateway/keeper/msg_server_update_param.go:34]: I chose not to rename it this time. My rationale is that it requires changing it in lots of other places which is not conducive to making the "adding parameters" step-by-step instructions any more concise. :sweat_smile: 

Do you have a strong opinion on this and/or a rationale that you could share?

File: x/gateway/keeper/msg_server_update_param_test.go (hunk 4 above)
Comment (bryanchriswhite) [reply x/gateway/keeper/msg_server_update_param_test.go:27]: :pray: Ty sir!

File: x/gateway/keeper/params_test.go (hunk 5 above)
Comment (bryanchriswhite) [reply x/gateway/keeper/params_test.go:24]: I expect this sort of copypasta to stop once I can start using the updated "adding params" doc. :sweat_smile: 

Thanks for catching these! :raised_hands:

File: tools/scripts/params/gateway_min_stake.json
Comment (bryanchriswhite) [inline tools/scripts/params/gateway_min_stake.json:1]: Forgot to add `gateway_all.json`.

File: proto/poktroll/gateway/params.proto [hunk 9]
```diff
@@ -5,10 +5,13 @@ option go_package = "github.com/pokt-network/poktroll/x/gateway/types";
...
 import "amino/amino.proto";
+import "cosmos/base/v1beta1/coin.proto";
 import "gogoproto/gogo.proto";
 
 // Params defines the parameters for the module.
 message Params {
   option (amino.name) = "poktroll/x/gateway/Params";
   option (gogoproto.equal) = true;
+
+  cosmos.base.v1beta1.Coin min_stake  = 1 [(gogoproto.jsontag) = "min_stake"];
```
Comment (bryanchriswhite) [inline proto/poktroll/gateway/params.proto:16]: Add godoc-style comment.

File: x/gateway/keeper/msg_server_update_param.go (hunk 3 above)
Comment (red-0ne) [reply x/gateway/keeper/msg_server_update_param.go:31]: Mostly because we're doing it everywhere else :smile:

**Review by red-0ne — APPROVED** [review]
Left a couple of comments but LGTM.

File: x/gateway/types/params.go [hunk 10]
```diff
@@ -1,30 +1,79 @@
...
+	if minStakeCoin.Denom != volatile.DenomuPOKT {
+		return ErrGatewayParamInvalid.Wrapf(
+			"invalid min_stake denom %q; expected %q",
+			minStakeCoin.Denom, volatile.DenomuPOKT,
+		)
+	}
+	if minStakeCoin.IsZero() || minStakeCoin.IsNegative() {
+		return ErrGatewayParamInvalid.Wrapf("invalid min stake amount must be greater than zero: %s", minStakeCoin)
+	}
+	if minStakeCoin.Denom != volatile.DenomuPOKT {
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/gateway/types/params.go:74]: Nice catch - thanks! :raised_hands:

**Pull request merged**